### PR TITLE
Add ipa_pwpolicy module

### PR DESCRIPTION
**Pull request by adralioh**
##### SUMMARY

Add a module for modifying FreeIPA password policies

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

ipa_pwpolicy

##### ADDITIONAL INFORMATION

The module is based heavily on the existing IPA modules, and extends the base IPA arguments/documentation

A couple considerations:
- The arguments all use the raw LDAP name, which are pretty long, eg `krbpwdfailurecountinterval`. I added shorter aliases based on the `ipa` command that are easier to use.

  I did it this way because this is how the existing IPA modules do it, but I can change the default name to the shorter versions if preferred.
- All of the arguments are strings instead of integers because they're passed directly to the IPA web API, which stores them as strings. I can switch them to integers if preferred.

### CODE REVIEW

File: plugins/modules/identity/ipa/ipa_pwpolicy.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+      ipa_host: ipa.example.com
+      ipa_user: admin
+      ipa_pass: topsecret
+'''
+
+RETURN = r'''
+pwpolicy:
+    description: Password policy as returned by IPA API
+    returned: always
+    type: dict
```
Comment (felixfontein) [inline plugins/modules/identity/ipa/ipa_pwpolicy.py:116]: A `sample:` value would be nice.

**Review by felixfontein — COMMENTED** [review]
Please note that all new modules and plugins are required to have tests (unit and/or integration tests).

**Comment by felixfontein** [comment]
Thanks for your PR! The module looks pretty polished. I'm not familiar with FreeIPA or the ipa_* modules, so there's not much I can say except that :)

File: plugins/modules/identity/ipa/ipa_pwpolicy.py (hunk 1 above)
Comment (adralioh) [reply plugins/modules/identity/ipa/ipa_pwpolicy.py:116]: Added in 92f88b266373cf6d9e4e71625b026776485e0d4e

**Comment by adralioh** [comment]
I added unit tests

ready_for_review

**Comment by felixfontein** [comment]
Cool, looks great! :)

@Akasurde @Nosmoht @fxfitz as maintainers of related modules, can you please take a look?

File: plugins/modules/identity/ipa/ipa_pwpolicy.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+short_description: Manage FreeIPA password policies
+description:
+- Add, modify, and delete password policy within IPA server.
+version_added: 1.3.0
+options:
+    cn:
+        description:
+        - Name of the group that the policy will apply to.
+        - If omitted, the global policy will be used.
+        aliases: ["group", "name"]
```
Comment (Andersson007) [inline plugins/modules/identity/ipa/ipa_pwpolicy.py:22]:
```suggestion
        aliases: ["group", "name"]
```
i would avoid aliases which is ok to add sometimes (but not initially), e.g. when a typo or something. So i'd choose one best option for every parameter.

**Review by Andersson007 — COMMENTED** [review]
@adralioh hi, thanks for the new module!
General things like the documentation, examples, etc. LGTM
would be good to get feedback from mentioned people

**Comment by Andersson007** [comment]
folks from freeipa team hinted that there's a special repo https://github.com/freeipa/ansible-freeipa/
and it looks like there's a similar module already implemented https://github.com/freeipa/ansible-freeipa/blob/master/plugins/modules/ipapwpolicy.py
@adralioh did you see that? Maybe better just to add missed functionality (if it is) to the existing one?

**Comment by Andersson007** [comment]
we're now discussing in IRC ansible-community the possibility of deprecating this stuff and redirecting people to the special repo

**Comment by adralioh** [comment]
The module doesn't do anything that the one in the `ansible-freeipa` repo doesn't do as far as I can tell. I wanted to add the module here because it was the only IPA functionality I needed in my playbooks that's currently missing from `community.general`.

Was there a consensus on whether or not to deprecate the IPA modules? If they're not going to be deprecated, I think it would still be worth adding this module.

File: plugins/modules/identity/ipa/ipa_pwpolicy.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+---
+module: ipa_pwpolicy
+author: Adralioh (@adralioh)
+short_description: Manage FreeIPA password policies
+description:
+- Add, modify, and delete password policy within IPA server.
+version_added: 1.3.0
```
Comment (felixfontein) [inline plugins/modules/identity/ipa/ipa_pwpolicy.py:16]:
```suggestion
version_added: 2.0.0
```

**Comment by felixfontein** [comment]
Since there hasn't really been any discussions, it might be better to add the module here for now. @Andersson007 what do you think?

**Comment by Andersson007** [comment]
> Since there hasn't really been any discussions, it might be better to add the module here for now. @Andersson007 what do you think?

Ok

File: plugins/modules/identity/ipa/ipa_pwpolicy.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    krbpwdfailurecountinterval:
+        description: Period after which lockout is enforced (seconds).
+        aliases: ["failinterval"]
+        type: str
+    krbpwdlockoutduration:
+        description: Period for which lockout is enforced (seconds).
+        aliases: ["lockouttime"]
+        type: str
+extends_documentation_fragment:
+- community.general.ipa.documentation
```
Comment (Andersson007) [inline plugins/modules/identity/ipa/ipa_pwpolicy.py:69]:
```suggestion
- community.general.ipa.documentation
notes:
- Supports C(check_mode).
```

File: plugins/modules/identity/ipa/ipa_pwpolicy.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+      group: sysops
+      state: absent
+      ipa_host: ipa.example.com
+      ipa_user: admin
+      ipa_pass: topsecret
+'''
+
+RETURN = r'''
+pwpolicy:
+    description: Password policy as returned by IPA API
```
Comment (Andersson007) [inline plugins/modules/identity/ipa/ipa_pwpolicy.py:114]:
```suggestion
    description: Password policy as returned by IPA API.
```

**Comment by Andersson007** [comment]
@adralioh thanks for the work! Let's continue here then.
my comment about aliases is still relevant.

File: plugins/modules/identity/ipa/ipa_pwpolicy.py (hunk 3 above)
Comment (adralioh) [reply plugins/modules/identity/ipa/ipa_pwpolicy.py]: Changed in cbfd9b1

File: plugins/modules/identity/ipa/ipa_pwpolicy.py (hunk 4 above)
Comment (adralioh) [reply plugins/modules/identity/ipa/ipa_pwpolicy.py:69]: Added in 09bca67

File: plugins/modules/identity/ipa/ipa_pwpolicy.py (hunk 5 above)
Comment (adralioh) [reply plugins/modules/identity/ipa/ipa_pwpolicy.py]: Added in ec7ee77

**Comment by adralioh** [comment]
@Andersson007 Regarding the aliases, I'm not sure which ones are the best option to keep.

The long names that I have as the default option name, eg `krbpwdfailurecountinterval`, are the raw LDAP name as returned by the IPA API. \
The shorter aliases, eg `failinterval`, are the names used on the commandline by the `ipa` command.

The other IPA modules mostly only use the LDAP name, but they're not as long as the ones in this module, so are more manageable.

Also, for reference, the module in the `ansible-freeipa` repo has aliases for both names.

I'm leaning towards removing the shorter aliases so it's consistent with the other modules, but I wanted to see what you thought.

File: plugins/modules/identity/ipa/ipa_pwpolicy.py [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+author: Adralioh (@adralioh)
+short_description: Manage FreeIPA password policies
+description:
+- Add, modify, and delete password policy within IPA server.
+version_added: 2.0.0
+options:
+    cn:
+        description:
+        - Name of the group that the policy will apply to.
+        - If omitted, the global policy will be used.
```
Comment (Andersson007) [inline plugins/modules/identity/ipa/ipa_pwpolicy.py:21]:
```suggestion
        - Name of the group that the policy will apply to.
        - If omitted, the global policy will be used.
```
Also try to use the present tense where applicable (it's from the development documentation as well)

**Comment by Andersson007** [comment]
@adralioh IMO we could use the short forms because as i can see they don't intersect with each other but we could put the long forms to descriptions, e.g. (very roughly):
```
    maxfile:
        description: Maximum password lifetime (in days). Corresponds to the C(krbmaxpwdlife) raw LDAP name.
```
BTW there's a typo in `maxfile`. I think `maxpwdlife` would sound better.

**Comment by Andersson007** [comment]
> @adralioh IMO we could use the short forms because as i can see they don't intersect with each other but we could put the long forms to descriptions, e.g. (very roughly):
> 
> ```
>     maxlife:
>         description: Maximum password lifetime (in days). Corresponds to the C(krbmaxpwdlife) raw LDAP name.
> ```
> 
> I think `maxpwdlife` would sound better.
> 

, providing that to know the raw names can be important for anyone. If not, i'd leave only the short forms

File: plugins/modules/identity/ipa/ipa_pwpolicy.py (hunk 6 above)
Comment (adralioh) [reply plugins/modules/identity/ipa/ipa_pwpolicy.py]: Changed in 8349dcd20a4c7b248eb3952a5dfed93c125e31b8

**Comment by adralioh** [comment]
@Andersson007 Thanks, I changed them to short names, and removed mention of the LDAP long names since they probably aren't important.

I renamed some of the short names to be more descriptive, such as `maxlife` to `maxpwdlife` like you mentioned, and `maxfail` to `maxfailcount`.

I also improved the description of some of the options so that they're better worded.

**Review by felixfontein — APPROVED** [review]

**Review by Andersson007 — APPROVED** [review]

**Comment by Andersson007** [comment]
@adralioh thanks for the new module!
@felixfontein thanks for reviewing!

**Comment by Andersson007** [comment]
@adralioh i'll add you as a maintainer to the special BOTMETA.yml file next week to keep you informed about related issues / PRs

**Pull request merged**